src/redux/reducers/drawing_canvas.ts

```diff
@@ -1,19 +1,70 @@
 import * as actions from "../actions/drawing_canvas.ts";
 import {Action} from "redux";
 import {AppState} from "../state.ts";
-import {AddShapeAction} from "../actions/drawing_canvas.ts";
+import {AddShapeAction, MoveShapeAction, SelectShapeAction, DeselectShapeAction, ChangeLevelShapeAction} from "../actions/drawing_canvas.ts";
 import {Shape} from "../../models.ts";
 
 export function drawingCanvasReducer(state: AppState, action: Action): AppState {
   switch (action.type) {
     case actions.ADD_SHAPE:
       return addShape(state, action as AddShapeAction);
+    case actions.MOVE_SHAPE:
+      return moveShape(state, action as MoveShapeAction);
+    case actions.SELECT_SHAPE:
+      return selectShape(state, action as SelectShapeAction);
+    case actions.DESELECT_SHAPE:
+      return deselectShape(state, action as DeselectShapeAction);
+    case actions.CHANGE_LEVEL_SHAPE:
+      return changeLevelShape(state, action as ChangeLevelShapeAction);
     default:
       return state;
   }
 }
 
 function addShape(state: AppState, action: AddShapeAction): AppState {
   const {offset, shape, color} = action;
-  return Object.assign({}, state, { shapes: [...state.shapes,  new shape(offset.x - 50, offset.y, 50, color)]});
+  return Object.assign({}, state, { shapes: [...state.shapes, new shape(offset.x - 50, offset.y, 50, color)] });
+}
+
+function moveShape(state: AppState, action: MoveShapeAction): AppState {
+  if (state.selectedShapeId) {
+    const {x, y} = action;
+    const newShapes = state.shapes.map((shape: Shape) => {
+      if (shape.id == state.selectedShapeId) {
+        shape.x = x;
+        shape.y = y
+      }
+      return shape;
+    });
+    return Object.assign({}, state, { shapes: [...newShapes] });
+  } else {
+    return state;
+  }
+}
```

Comment:
Instead of grabbing the selected shape from the state, you could pass it along with the action. This would give you a bit more reusability, as you might want to reuse this logic for moving an item that isn't selected. In fact, this function could be generalized to:

```ts
function moveShape(shapes: Shape[], shapeId: string, x: number, y: number): Shape[] {
  return state.shapes.map((shape: Shape) => {
    if (shape.id === shapeId) {
      shape.x = x;
      shape.y = y
    }
    return shape;
  });
}
```